### ISSUE TITLE
`multiple_crate_versions` false negative in source repo but not as dependency

### ISSUE THREAD

**Issue by spearman** [rust-lang/rust-clippy#15643]
### Summary

In the crate [nsys-enet](https://crates.io/crates/nsys-enet) I have:
```
[lints.clippy]
multiple-crate-versions = "warn"
```
And because of a dependency, both `"num-traits 0.1.43"` and `"num-traits 0.2.19"` are in the dependency graph. (I will probably try replacing the dependency that causes this, however I have run into this same issue with other crates and this one is the most minimal).

However when running `cargo clippy` no warning is emitted. But if I create a new crate and add `nsys-enet` as a dependency and enable the `multiple_crate_versions` lint, then a warning is emitted:
```
warning: multiple versions for dependency `num-traits`: 0.1.43, 0.2.19
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions
  = note: requested on the command line with `-W clippy::multiple-crate-versions`
```

I also encounter the same issue with the [nsys-gl-utils](https://crates.io/crates/nsys-gl-utils) crate, with a larger number of dependencies.

Version info:
```
rustc 1.91.0-nightly (9c27f27ea 2025-09-08)
cargo 1.91.0-nightly (761c4658d 2025-09-04)
clippy 0.1.91 (9c27f27ea3 2025-09-08)
```

### Lint Name

multiple_crate_versions

### Reproducer

I tried this code:

```
git clone https://github.com/spearman/enet-rs && cd enet-rs && cargo clippy
```

I expected to see this happen:

```
warning: multiple versions for dependency `num-traits`: 0.1.43, 0.2.19
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions
  = note: requested on the command line with `-W clippy::multiple-crate-versions`
```

Instead, this happened:

[nothing]

### Version

```text
rustc 1.91.0-nightly (9c27f27ea 2025-09-08)
binary: rustc
commit-hash: 9c27f27ea3bab79a2fec827ef3ae0009959d60f4
commit-date: 2025-09-08
host: x86_64-unknown-linux-gnu
release: 1.91.0-nightly
LLVM version: 21.1.0
```

**Comment by KisaragiEffective** [comment]
I doubted that's just a build-dependencies, but these are not.
`num-traits@0.1.43` re-exports `num-traits@0.2.19` - so the `0.1.43` depends on `0.2.19` by  mean of normal dependency.
